File: bilibili/src/main/java/com/bilibili/ui/bangumi/BangumiPresenter.java
package com.bilibili.ui.bangumi;

import android.util.Log;

import com.bilibili.model.api.ApiHelper;
import com.bilibili.model.api.BangumiApis;
import com.bilibili.model.bean.ResultListResponse;
import com.bilibili.model.bean.ResultObjectResponse;
import com.bilibili.model.bean.bangumi.BangumiIndexFall;
import com.bilibili.model.bean.bangumi.BangumiIndexPage;
import com.bilibili.ui.bangumi.viewbinder.BangumiDividerBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiHomeBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexFollowBinder;
import com.bilibili.ui.bangumi.viewbinder.BangumiIndexRecommendBinder;
import com.common.base.AbsBasePresenter;
import com.common.util.DateUtil;

import javax.inject.Inject;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import me.drakeet.multitype.Items;

/**
 * Created by miserydx on 17/6/29.
 */

public class BangumiPresenter extends AbsBasePresenter<BangumiContract.View> implements BangumiContract.Presenter {

    private static final String TAG = BangumiPresenter.class.getSimpleName();
    public static final int STATE_NORMAL = 0;
    public static final int STATE_INITIAL = 1;
    public static final int STATE_REFRESHING = 2;
    public static final int STATE_LOAD_MORE = 3;

    private BangumiApis bangumiApis;
    private long cursor = 0;

    @Inject
    public BangumiPresenter(BangumiApis bangumiApis) {
        this.bangumiApis = bangumiApis;
    }

    @Override
    public void loadData() {
        getIndexPage(STATE_INITIAL);
    }

    @Override
    public void pullToRefresh() {
        getIndexPage(STATE_REFRESHING);
    }

    private void getIndexPage(final int state) {
        bangumiApis.getIndexPage(ApiHelper.APP_KEY,
                ApiHelper.BUILD,
                ApiHelper.MOBI_APP,
                ApiHelper.PLATFORM,
                DateUtil.getSystemTime())
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.newThread())
                .map(new Function<ResultObjectResponse<BangumiIndexPage>, Items>() {
                    @Override
                    public Items apply(ResultObjectResponse<BangumiIndexPage> bangumiIndexPageRes) throws Exception {
                        return getItems(bangumiIndexPageRes.getData());
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Items>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        registerRx(d);
                        mView.onRefreshingStateChanged(true);
                    }

                    @Override
                    public void onNext(Items items) {
                        mView.onDataUpdated(items, state);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, "onError");
                        e.printStackTrace();
                        mView.showLoadFailed();
                        mView.onRefreshingStateChanged(false);
                    }

                    @Override
                    public void onComplete() {
                        mView.onRefreshingStateChanged(false);
                    }
                });
    }

    private Items getItems(BangumiIndexPage bangumiIndexPage) {
        Items items = new Items();
        items.add(new BangumiIndexFollowBinder.BangumiIndexFollow());
        items.add(new BangumiHomeBinder.BangumiHome());
        items.add(new BangumiDividerBinder.BangumiDivider());
        items.add(new BangumiIndexRecommendBinder.BangumiIndexRecommend(BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_JP_RECOMMEND));
        items.addAll(bangumiIndexPage.getRecommend_jp().getRecommend());
        items.add(bangumiIndexPage.getRecommend_jp().getFoot().get(0));
        items.add(new BangumiDividerBinder.BangumiDivider());
        items.add(new BangumiIndexRecommendBinder.BangumiIndexRecommend(BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_CN_RECOMMEND));
        items.addAll(bangumiIndexPage.getRecommend_cn().getRecommend());
        items.add(bangumiIndexPage.getRecommend_cn().getFoot().get(0));
        return items;
    }

    @Override
    public void loadMore() {
        bangumiApis.getIndexFall(ApiHelper.APP_KEY,
                ApiHelper.BUILD, cursor,
                ApiHelper.MOBI_APP,
                ApiHelper.PLATFORM,
                DateUtil.getSystemTime())
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.newThread())
                .map(new Function<ResultListResponse<BangumiIndexFall>, Items>() {
                    @Override
                    public Items apply(ResultListResponse<BangumiIndexFall> bangumiIndexFallRes) throws Exception {
                        Items items = new Items();
                        if (cursor == 0L) {
                            items.add(new BangumiDividerBinder.BangumiDivider());
                            items.add(new BangumiIndexRecommendBinder.BangumiIndexRecommend(
                                    BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_EDITORS_RECOMMEND));
                        }
                        for (BangumiIndexFall bangumiIndexFall : bangumiIndexFallRes.getData()) {
                            if (bangumiIndexFall.getCursor() != 0.0) {
                                cursor = (long) (bangumiIndexFall.getCursor());
                            }
                            items.add(bangumiIndexFall);
                        }
                        return items;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<Items>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        registerRx(d);
                    }

                    @Override
                    public void onNext(Items items) {
                        mView.onDataUpdated(items, STATE_LOAD_MORE);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(BangumiFragment.TAG, "onError");
                        e.printStackTrace();
                        mView.showLoadMoreError();
                    }

                    @Override
                    public void onComplete() {

                    }
                });

    }

    @Override
    public void releaseData() {

    }
}

File: bilibili/src/main/java/com/bilibili/ui/bangumi/BangumiPresenter.kt
package com.bilibili.ui.bangumi

import android.util.Log

import com.bilibili.model.api.ApiHelper
import com.bilibili.model.api.BangumiApis
import com.bilibili.model.bean.BangumiIndexPage
import com.bilibili.ui.bangumi.viewbinder.*
import com.common.base.AbsBasePresenter
import com.common.util.DateUtil

import javax.inject.Inject

import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import me.drakeet.multitype.Items

/**
 * Created by miserydx on 17/6/29.
 */

class BangumiPresenter @Inject
constructor(private val bangumiApis: BangumiApis) : AbsBasePresenter<BangumiContract.View>(), BangumiContract.Presenter {

    companion object {
        private val TAG = BangumiPresenter::class.java.simpleName
        val STATE_NORMAL = 0
        val STATE_INITIAL = 1
        val STATE_REFRESHING = 2
        val STATE_LOAD_MORE = 3
        val STATE_LOAD_ERROR = 4
    }

    private var cursor: Long = 0

    override fun loadData() {
        getIndexPage(STATE_INITIAL)
    }

    override fun pullToRefresh() {
        getIndexPage(STATE_REFRESHING)
    }

    private fun getIndexPage(state: Int) {
        bangumiApis.getIndexPage(ApiHelper.APP_KEY, ApiHelper.BUILD, ApiHelper.MOBI_APP, ApiHelper.PLATFORM, DateUtil.getSystemTime())
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.newThread())
                .map { bangumiIndexPageRes -> getItems(bangumiIndexPageRes.data) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<Items> {
                    override fun onSubscribe(@NonNull d: Disposable) {
                        registerRx(d)
                        mView.onRefreshingStateChanged(true)
                    }

                    override fun onNext(@NonNull items: Items) {
                        mView.onDataUpdated(items, state)
                    }

                    override fun onError(@NonNull e: Throwable) {
                        Log.e(TAG, "onError")
                        e.printStackTrace()
                    }

                    override fun onComplete() {
                        mView.onRefreshingStateChanged(false)
                    }
                })
    }

    private fun getItems(bangumiIndexPage: BangumiIndexPage): Items {
        val items = Items()
        items.add(BangumiIndexFollowBinder.BangumiIndexFollow())
        items.add(BangumiHomeBinder.BangumiHome())
        items.add(BangumiDividerBinder.BangumiDivider())
        items.add(BangumiIndexRecommendBinder.BangumiIndexRecommend(BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_JP_RECOMMEND))
        for (recommend in bangumiIndexPage.recommend_jp.recommend) {
            items.add(recommend)
        }
        items.add(bangumiIndexPage.recommend_jp.foot[0])
        items.add(BangumiDividerBinder.BangumiDivider())
        items.add(BangumiIndexRecommendBinder.BangumiIndexRecommend(BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_CN_RECOMMEND))
        for (recommend in bangumiIndexPage.recommend_cn.recommend) {
            items.add(recommend)
        }
        items.add(bangumiIndexPage.recommend_cn.foot[0])
        return items
    }

    override fun loadMore() {
        bangumiApis.getIndexFall(ApiHelper.APP_KEY, ApiHelper.BUILD, cursor, ApiHelper.MOBI_APP, ApiHelper.PLATFORM, DateUtil.getSystemTime())
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.newThread())
                .map { bangumiIndexFallRes ->
                    val items = Items()
                    if (cursor == 0L) {
                        items.add(BangumiDividerBinder.BangumiDivider())
                        items.add(BangumiIndexRecommendBinder.BangumiIndexRecommend(BangumiIndexRecommendBinder.BangumiIndexRecommend.SECTION_EDITORS_RECOMMEND))
                    }
                    for (bangumiIndexFall in bangumiIndexFallRes.data) {
                        if (bangumiIndexFall.cursor != 0.0) {
                            cursor = bangumiIndexFall.cursor.toLong()
                        }
                        items.add(bangumiIndexFall)
                    }
                    items
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<Items> {
                    override fun onSubscribe(@NonNull d: Disposable) {
                        registerRx(d)
                    }

                    override fun onNext(@NonNull items: Items) {
                        mView.onDataUpdated(items, STATE_LOAD_MORE)
                    }

                    override fun onError(@NonNull e: Throwable) {
                        Log.e(TAG, "onError")
                        e.printStackTrace()
                        mView.onDataUpdated(null, STATE_LOAD_ERROR)
                    }

                    override fun onComplete() {}
                })
    }

    override fun releaseData() {

    }
}
